### ISSUE TITLE
`nn.DataPallel` fails for checkpoint feature

### ISSUE THREAD

**Issue by wandering007** [gpleiss/efficient_densenet_pytorch#36]
I just want to benchmark the new implementation of efficient densenet with the code [here](https://gist.github.com/wandering007/d7f3db11285f382fefa39a9cda99ad0d). However, it seems that the used checkpointed modules are not broadcast to multiple GPUs as I got the following errors:

```
  File "/home/changmao/efficient_densenet_pytorch/models/densenet.py", line 16, in bn_function
    bottleneck_output = conv(relu(norm(concated_features)))
  File "/home/changmao/anaconda3/lib/python3.6/site-packages/torch/nn/modules/module.py", line 491, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/changmao/anaconda3/lib/python3.6/site-packages/torch/nn/modules/batchnorm.py", line 49, in forward
    self.training or not self.track_running_stats, self.momentum, self.eps)
  File "/home/changmao/anaconda3/lib/python3.6/site-packages/torch/nn/functional.py", line 1194, in batch_norm
    training, momentum, eps, torch.backends.cudnn.enabled
RuntimeError: Expected tensor for argument #1 'input' to have the same device as tensor for argument #2 'weight'; but device 1 does not equal 0 (while checking arguments for cudnn_batch_norm)
```
I was confused and couldn't figure it out. And it seems that the backward of `nn.DataParallel` is too slow (when I experienced no errors before, weird).

**Comment by gpleiss** [comment]
Oooh @wandering007 good catch. I'll take a look.

**Comment by wandering007** [comment]
@gpleiss This implementation (https://github.com/wandering007/efficient-densenet-pytorch) has good support for `nn.DataParallel`, which may be helpful.

**Comment by gpleiss** [comment]
Just merged in #39 . @wandering007 , can you confirm that this fixes the issue?

**Comment by wandering007** [comment]
@gpleiss Yes, it works fine. However, there is one thing that I've noticed before and have to mention, though out of the scope of this issue. With checkpointing feature, the whole autograd computation graph is broken into pieces. And the current `nn.DataParallel` backward process is roughly *backward in each GPU asynchronously* and *inter-GPU communication for collecting weight gradients (aka, reduce)* for **each** autograd computation graph. That is, if one checkpoint contains weight for updating, there would be an inter-gpu synchronization process for accumulating the gradients in it, which is time-consuming. Considering that the current efficient DenseNet contains so many checkpointed `nn.BatchNorm2d` modules, a lot of time will be spent on the inter-GPU communications and gradient accumulation. From my test, the backward of efficent DenseNet for multi-GPUs is at least 100x slower that the normal version...

**Comment by gpleiss** [comment]
@wandering007 hmmm that is problematic...

In general, I think that the checkpointing-based approach is probably what we should be doing moving forward. The original version was using some low-level calls which are no longer available in PyTorch. Using those low-level calls would require some C code, which is in my opinion undesirable for this package.

However, it sounds like the checkpointing-based code is practically unusable for the multi-GPU scenario. It's probably worthwhile bringing up an issue in the PyTorch repo about this. I'll see if there's a better solution in the meantime.

**Comment by wandering007** [comment]
@gpleiss It may be tough for now...To be frank, I am still in favor of the previous implementation (v0.3.1) via `_EfficientDensenetBottleneck` class and `_DummyBackwardHookFn` function without touching any C code. I've just made some improvements on it and it seems very neat and workable for PyTorch v0.4. You can check https://github.com/wandering007/efficient-densenet-pytorch/blob/master/efficient_utils.py if you are interested.

**Comment by yzcjtr** [comment]
Maybe this issue could have been made more clear in the readme. I followed the implementation in my project but found it doesn't work with dataparallel ...

**Comment by gpleiss** [comment]
@theonegis can you provide more information? What version of PyTorch, what OS, what version of CUDA, what GPUs, etc.? Also, could you open up a new issue for this?

**Comment by theonegis** [comment]
@gpleiss I have opened a new issues. [Segmentation fault (core dumped) error for multiple GPUs](https://github.com/gpleiss/efficient_densenet_pytorch/issues/47).
Thanks a lot.

**Comment by yzcjtr** [comment]
Hi @gpleiss , really sorry for my previous misunderstanding. I'm confronted with a similar situation as @theonegis . I will provide more information in his new issue. Thanks.

**Comment by csrhddlam** [comment]
The PyTorch official checkpointing is slow on MultiGPUs as explained by @wandering007 . https://github.com/csrhddlam/pytorch-checkpoint solves this issue.
